README.md
About
=====

This is the optional pattern designed to disaster recovery which monitoring provides to platform less patterns.
Currently supported oparationg system:

* CentOS (6.7 7.2)

Currently supported monitoring software:

* Zabbix (2.2 3.0)

For more information, please visit [official web site](http://cloudconductor.org/).

Requirements
============

Prerequisites
-------------

- cloudconductor (>= 2.0)
- any platform pattern

How to use patterns
============

You can apply this pattern by using CloudConductor CLI tools or REST API.
Please see [Getting started](http://cloudconductor.org/en/documents/getting-started) in official web site to know
how the parameters you input to CloudConductor CLI tools or REST API are converted into
chef attributes.

Attributes
==========

The attributes not described here have default values, and you can change the value of them if you need.
Please see the attribute files if you want to know what kind of attributes are in this pattern.

- Attribute

| Attribute             | Value                                           |
| -------------------- |:------------------------------------------------ |
| cloudconductor_url   | Cloudconductor grobal address (port)             |
| cloudconductor_auth_token | Cloudconductor accsess secret key                |
| zabbix_url           | zabbix-web front address                         |
| zabbix_user          | zabbix usertype: Zabbix Admin and RW Permissions |
| zabbix_password      | zabbix password                                  |
| zabbix_template      | Attach Zabbix template to created host           |

sample) config/environment_attribute.json.smp

```config/environment_attribute.json.smp
{
  "system_monitoring_pattern": {
    "common": {
      "cloudconductor_url": "[[CLOUDCONDUCTOR_URL[:8080]]]",
      "cloudconductor_auth_token: "[[CLOUDCONDUCTOR_TOKEN]]"
    },
    "zabbix_part": {
      "zabbix_url": "http://[[ZABBIX.ADDRESS]]/zabbix",
      "zabbix_user": "[[USERNAME]]",
      "zabbix_password": "[[PASSWORD]]",
      "zabbix_template": "Template App HTTP Service"
    }
  }
}
```

- How to

Please be sure to specify the domain when you create a system
  `cc-cli system create --project sample_project --name dr-pattern --description "disaster recovery pattern" --domain "disaster.sampledomain.com"`

  `cc-cli environment create --system sample_system --blueprint sample_blueprint --clouds cloud-aws --name sample_environment --version 1 --user-attribute-file ~/environment_attribute.json`


Copyright and License
=====================

Copyright 2016 TIS inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


Contact
========

For more information: <http://cloudconductor.org/>

Report issues and requests: <https://github.com/cloudconductor-patterns/system_monitoring_pattern/issues>

Send feedback to: <ccndctr@gmail.com>


CHANGELOG.md
CHANGELOG
=========

## version 1.0.0 (2016/03/30)

  - First release of system monitoring pattern that provides user with DR (disaster recovery) feature.


spec/zabbix_spec.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Copyright 2014-2015 TIS Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import os
import logging
import urllib2
import json
import base64
from zabbix.api import ZabbixAPI
from distutils.version import StrictVersion

def ret_url_encode(encode_str):

    return encode_str.replace('+','%2B')

def get_kv(keystrings,token):

    import urllib2

    response = urllib2.urlopen('http://127.0.0.1:8500/v1/kv/' + keystrings + '?token=' + token)

    return response.read()

if __name__ == '__main__':

    import sys

    consul_kv = get_kv("cloudconductor/patterns/system_monitoring_pattern/attributes",ret_url_encode(os.environ['CONSUL_SECRET_KEY']))
    decode_consul_kv = json.loads(consul_kv)
    decode_user_attribute = json.loads(base64.b64decode(decode_consul_kv[0]["Value"]))
    zabbix_user = decode_user_attribute['zabbix_part']['zabbix_user']
    zabbix_password = decode_user_attribute['zabbix_part']['zabbix_password']
    zabbix_url = decode_user_attribute['zabbix_part']['zabbix_url']
    zabbix_template = decode_user_attribute['zabbix_part']['zabbix_template']

    # Getting Consul KV
    consul_kv = get_kv("cloudconductor/system_domain",ret_url_encode(os.environ['CONSUL_SECRET_KEY']))
    decode_consul_kv = json.loads(consul_kv)
    decode_user_attribute = json.loads(base64.b64decode(decode_consul_kv[0]["Value"]))
    system_domain = decode_user_attribute['dns']
    system_name = decode_user_attribute['name']

    # Create ZabbixAPI class instance
    try:
        zapi = ZabbixAPI(url=zabbix_url, user=zabbix_user, password=zabbix_password)
    except Exception, e:
        print e, 'ZabbixAPI: Authenticate failed.'

    # Get Zabbix API version
    try:
        result_version = zapi.api_version()
    except:
        result_version = "2.4"

    # Hostgroup get
    hostgroup = system_name
    try:
        result_hostgroup = zapi.do_request('hostgroup.get', { 'filter': {'name': hostgroup}})
        if not result_hostgroup['result']:
            print 'Can not be found host group ' + hostgroup
            exit (-1)
        else:
            hostgroup_id = result_hostgroup['result'][0]['groupid']
    except Exception, e:
        print e, 'ZabbixAPI: hostgroup.get exist failed.'

    # Host get
    hostname = system_domain
    try:
        result_host = zapi.do_request('host.get', { 'filter': {'name': hostname}})
        if not result_host['result']:
            print 'Can not be found host ' + hostname
            exit (-1)
    except Exception, e:
        print e, 'ZabbixAPI: host.get exist failed.'

    # Action registry
    action = 'FailOver_' + hostname
    try:
        result_action = zapi.do_request('action.get', { 'filter': {'name': action}})
        if not result_action['result']:
            print 'Can not be found action ' + action
            exit (-1)
    except Exception, e:
        print e, 'ZabbixAPI: action.get exist failed.'

    exit ()


registory/zabbix_registory.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Copyright 2014-2015 TIS Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import os
import logging
import urllib2
import json
import base64
from zabbix.api import ZabbixAPI
from distutils.version import StrictVersion

def ret_url_encode(encode_str):

    return encode_str.replace('+', '%2B')

def get_kv(keystrings, token):

    import urllib2

    response = urllib2.urlopen('http://127.0.0.1:8500/v1/kv/' + keystrings + '?token=' + token)

    return response.read()

def load_yaml(yaml_file):
    import yaml

    with open('config/' + yaml_file) as reader:
        return yaml.load(reader.read())

def host_create_params(hostname, hostgroupid, template_id):

    ret_parameters = {
        'host': hostname,
        'interfaces': [
            {
                'type': 1,
                'main': 1,
                'ip': '',
                'dns': hostname,
                'port': 10050,
                'useip': 0
            }
        ],
        'groups': [
            {
                'groupid': hostgroup_id
            }
        ],
        'templates': [
            {
                'templateid': template_id
            }
        ]
    }
    return ret_parameters

def action_create_params(action, host_id, operation, version):

    if StrictVersion(version) >= StrictVersion("2.4"):
        ret_parameters = {
            'name': action,
            'eventsource': 0,
            'status': 0, # enabled
            'esc_period': 120,
            'def_shortdata': '{TRIGGER.NAME}: {TRIGGER.STATUS}',
            'def_longdata': '{TRIGGER.NAME}: {TRIGGER.STATUS}\r\nLast value: {ITEM.LASTVALUE}\r\n\r\n{TRIGGER.URL}',
            'operations': generate_operations(operation),
            'filter': {
                'evaltype': 1, # AND
                'conditions': generate_conditions(host_id)
            }
    }
    else:
        ret_parameters = {
            'name': action,
            'eventsource': 0,
            'status': 0, # enabled
            'esc_period': 120,
            'def_shortdata': '{TRIGGER.NAME}: {TRIGGER.STATUS}',
            'def_longdata': '{TRIGGER.NAME}: {TRIGGER.STATUS}\r\nLast value: {ITEM.LASTVALUE}\r\n\r\n{TRIGGER.URL}',
            'operations': generate_operations(operation),
            'evaltype': 1, # AND
            'conditions': generate_conditions(host_id)
        }

    return ret_parameters

def action_update_params(action_id, host_id, operation, version):

    if StrictVersion(version) >= StrictVersion("2.4"):
        ret_parameters = {
            'actionid': action_id,
            'operations': generate_operations(operation),
            'filter': {
                'conditions': generate_conditions(host_id)
            }
    }
    else:
        ret_parameters = {
            'actionid': action_id,
            'operations': generate_operations(operation),
            'conditions': generate_conditions(host_id)
        }

    return ret_parameters

def generate_conditions(host_id):

    ret_parameters = [
        {
            'conditiontype': '1', # host
            'operator': '0', # equal
            'value': host_id
        },
        {
            'conditiontype': '5', # trigger value
            'operator': '0', # equal
            'value': '1'
        }
    ]

    return ret_parameters

def generate_operations(operation):

    ret_parameters = [
        {
            "operationtype": "1", # remote command
            "opcommand_hst": {
                "hostid": "0"
            },
            "opcommand": {
                "type": "0", # custom script
                "command": operation,
                "execute_on": "1" # Zabbix server
            }
        }
    ]

    return ret_parameters

def operation(environment_id, auth_token, url):

    parameter = '\'{"switch": "true", "auth_token": "' + auth_token + '"}\''
    ret_parameters = 'curl -H "Content-Type:application/json" -X POST -d ' + parameter + ' ' + url.rstrip('/') + '/api/v1/environments/' + str(environment_id) + '/rebuild'

    return ret_parameters

if __name__ == '__main__':

    import sys

    # Settings Yaml
    # ex. yaml_data = load_yaml('monitoring.yml')

    # Settings Json
    consul_kv = get_kv("cloudconductor/patterns/system_monitoring_pattern/attributes", ret_url_encode(os.environ['CONSUL_SECRET_KEY']))
    decode_consul_kv = json.loads(consul_kv)
    decode_user_attribute = json.loads(base64.b64decode(decode_consul_kv[0]["Value"]))
    cloudconductor_url = decode_user_attribute['common']['cloudconductor_url']
    cloudconductor_token = decode_user_attribute['common']['cloudconductor_auth_token']
    zabbix_user = decode_user_attribute['zabbix_part']['zabbix_user']
    zabbix_password = decode_user_attribute['zabbix_part']['zabbix_password']
    zabbix_url = decode_user_attribute['zabbix_part']['zabbix_url']
    zabbix_template = decode_user_attribute['zabbix_part']['zabbix_template']

    # Getting Consul KV
    consul_kv = get_kv("cloudconductor/environment_id", ret_url_encode(os.environ['CONSUL_SECRET_KEY']))
    decode_consul_kv = json.loads(consul_kv)
    decode_user_attribute = json.loads(base64.b64decode(decode_consul_kv[0]["Value"]))
    environment_id = decode_user_attribute['environment_id']

    consul_kv = get_kv("cloudconductor/system_domain", ret_url_encode(os.environ['CONSUL_SECRET_KEY']))
    decode_consul_kv = json.loads(consul_kv)
    decode_user_attribute = json.loads(base64.b64decode(decode_consul_kv[0]["Value"]))

    try:
        system_domain = decode_user_attribute['dns']
    except Exception as w:
        system_domain = 'localhost'
        print w, 'Please dns name change the settings as soon as possible.'

    system_name = decode_user_attribute['name']

    # Create ZabbixAPI class instance
    try:
        zapi = ZabbixAPI(url=zabbix_url, user=zabbix_user, password=zabbix_password)
    except Exception as e:
        print e, 'ZabbixAPI: Authenticate failed.'
        exit (-1)

    # Get Zabbix API version
    try:
        result_version = zapi.api_version()
    except:
        result_version = "2.4"

    # Hostgroup create
    hostgroup = system_name
    try:
        result_hostgroup = zapi.do_request('hostgroup.get', { 'filter': {'name': hostgroup}})
        if not result_hostgroup['result']:
            result_hostgroup = zapi.do_request('hostgroup.create', {'name': hostgroup})
            hostgroup_id = result_hostgroup['result']['groupids'][0]
        else:
            hostgroup_id = result_hostgroup['result'][0]['groupid']
    except Exception as e:
        print e, 'ZabbixAPI: Hostgroup exist failed.'
        exit (-1)

    # Host create
    hostname = system_domain
    try:
        result_host = zapi.do_request('host.get', { 'filter': {'name': hostname}})
        result_template = zapi.do_request('template.get', { 'filter': {'host': zabbix_template}})
        if not result_host['result']:
            result_host = zapi.do_request('host.create', host_create_params(hostname, hostgroup_id, result_template['result'][0]['templateid']))
            host_id = result_host['result']['hostids'][0]
        else:
            host_id = result_host['result'][0]['hostid']
    except Exception as e:
        print e, 'ZabbixAPI: Host exist failed.'
        exit (-1)

    # Action registry
    action = 'FailOver_' + hostname
    try:
        result_action = zapi.do_request('action.get', { 'filter': {'name': action}})
        if str(result_action['result']) == '[]':
            result_action = zapi.do_request('action.create', action_create_params(action, host_id, operation(environment_id, cloudconductor_token, cloudconductor_url), result_version))
        else:
            result_action = zapi.do_request('action.update', action_update_params(result_action['result'][0]['actionid'], host_id, operation(environment_id, cloudconductor_token, cloudconductor_url), result_version))

    except Exception as e:
        print e, 'ZabbixAPI: Action exist failed.'
        exit (-1)

    exit (0)
